src/build.cc

```diff
@@ -783,16 +796,28 @@ Node* Builder::AddTarget(const string& name, string* err) {
 }
 
 bool Builder::AddTarget(Node* target, string* err) {
-  if (!scan_.RecomputeDirty(target, err))
+  std::vector<Node*> validation_nodes;
+  if (!scan_.RecomputeDirty(target, &validation_nodes, err))
     return false;
 
   if (Edge* in_edge = target->in_edge()) {
-    if (in_edge->outputs_ready())
-      return true;  // Nothing to do.
+    if (!in_edge->outputs_ready() &&
+        !plan_.AddTarget(target, err)) {
+      return false;
+    }
   }
 
-  if (!plan_.AddTarget(target, err))
-    return false;
```

Comment:
The inversion of the `outputs_ready` condition and return value look like refactoring in preparation for adding the validation nodes below.  Please split that refactoring into a preceding commit (still in this PR) with its own explanation.

Reply:
Done.